sdk/samples/iot/hub/src/paho_iot_hub_c2d_example.c

```diff
@@ -149,35 +188,62 @@ static az_result read_configuration_and_init_client()
   return AZ_OK;
 }
 
-static int on_received(void* context, char* topicName, int topicLen, MQTTClient_message* message)
+// Read OS environment variables using stdlib function
+static az_result read_configuration_entry(
+    const char* name,
+    const char* env_name,
+    char* default_value,
+    bool hide_value,
+    az_span buffer,
+    az_span* out_value)
 {
-  (void)context;
+  printf("%s = ", name);
+  char* env = getenv(env_name);
 
-  if (topicLen == 0)
+  if (env != NULL)
   {
-    // The length of the topic if there are one or more NULL characters embedded in topicName,
-    // otherwise topicLen is 0.
-    topicLen = (int)strlen(topicName);
+    printf("%s\n", hide_value ? "***" : env);
+    az_span env_span = az_span_from_str(env);
+    AZ_RETURN_IF_NOT_ENOUGH_SIZE(buffer, az_span_size(env_span));
+    az_span_copy(buffer, env_span);
+    *out_value = az_span_slice(buffer, 0, az_span_size(env_span));
+  }
+  else if (default_value != NULL)
+  {
+    printf("%s\n", default_value);
```

Comment:
this looks like dup code

if (env == NULL && default_value != NULL)
{
   env = default_value;
}

Reply:
almost.  one uses env_value, while the other uses default_value.  But it's unclear to me why sometimes the default value passed in is NULL and other times it's an empty string literal. ... We don't have a legit default value in either case, so this param doesn't seem useful.  If the environment val is NULL, I think we should just make the user set it.

Reply:
In some cases the default value can be useful as a string. DPS is a good example. There is a default endpoint for DPS that the user can overwrite if they would like. But we pass the default one in case they don't care to change it.

In the case that we require someone to supply something, the default value should be `NULL`. If it is optional then an empty string literal. If I am reading correctly then the `device_id` and `hostname` should be changed to `NULL` and not an empty string. The device cert should remain a `NULL` and the trusted cert should remain an empty string.

That's another thing on the list of things to do is to consolidate the reused code like "getting environment variables". We should look to merge these into a single `.c` since we have five similar versions which could grow out of sync.

Reply:
yay, i already started changing the device_id and hostname to NULL before I saw your comment!

Reply:
Yes, there is a lot of duplicated code between the samples.  Perfect time for a class structure... if this were C++. I will settle for a C solution :)

Reply:
My issue was not dup code across samples, but dup code *in* the samples